sway/desktop/transaction.c

```diff
@@ -0,0 +1,398 @@
+		// There is another transaction in progress - we must add this one to
+		// the queue so we complete after it.
+		struct sway_transaction *tail = server.head_transaction;
+		while (tail->next) {
+			tail = tail->next;
+		}
+		tail->next = transaction;
+	} else if (transaction->num_waiting) {
+		// There are no other transactions, but we're not applying immediately
+		// so we must jump in the queue so others will queue behind us.
+		server.head_transaction = transaction;
+	} else {
+		// There are no other transactions in progress, and this one has nothing
+		// to wait for, so we can skip the queue.
+		wlr_log(L_DEBUG, "Transaction %p has nothing to wait for", transaction);
+		transaction_apply(transaction);
+		transaction_destroy(transaction);
+		return;
+	}
+
+	if (transaction->num_waiting) {
+		// Set up a timer which the views must respond within
+		transaction->timer = wl_event_loop_add_timer(server.wl_event_loop,
+				handle_timeout, transaction);
+		wl_event_source_timer_update(transaction->timer, TIMEOUT_MS);
+	}
+
+	// The debug tree shows the pending/live tree. Here is a good place to
+	// update it, because we make a transaction every time we change the pending
+	// tree.
+	update_debug_tree();
+}
+
+static void set_instruction_ready(
+		struct sway_transaction_instruction *instruction) {
+	instruction->ready = true;
+	struct sway_transaction *transaction = instruction->transaction;
+
+	if (server.debug_txn_timings) {
+		struct timespec now;
+		clock_gettime(CLOCK_MONOTONIC, &now);
+		struct timespec *start = &transaction->commit_time;
+		float ms = (now.tv_sec - start->tv_sec) * 1000 +
+			(now.tv_nsec - start->tv_nsec) / 1000000.0;
+		wlr_log(L_DEBUG, "Transaction %p: %li/%li ready in %.1fms (%s)",
+				transaction,
+				transaction->num_configures - transaction->num_waiting + 1,
+				transaction->num_configures, ms,
+				instruction->container->name);
+
+	}
+
+	// If all views are ready, apply the transaction.
+	// If the transaction has timed out then its num_waiting will be 0 already.
+	if (transaction->num_waiting > 0 && --transaction->num_waiting == 0) {
+#if !TRANSACTION_DEBUG
+		wlr_log(L_DEBUG, "Transaction %p is ready", transaction);
+		wl_event_source_timer_update(transaction->timer, 0);
+		transaction_progress_queue();
+#endif
+	}
+}
+
+void transaction_notify_view_ready(struct sway_view *view, uint32_t serial) {
+	for (int i = 0; i < view->swayc->instructions->length; ++i) {
+		struct sway_transaction_instruction *instruction =
+			view->swayc->instructions->items[i];
+		if (instruction->serial == serial && !instruction->ready) {
+			set_instruction_ready(instruction);
+			return;
+		}
+	}
+}
+
+void transaction_notify_view_ready_by_size(struct sway_view *view,
+		int width, int height) {
+	for (int i = 0; i < view->swayc->instructions->length; ++i) {
+		struct sway_transaction_instruction *instruction =
+			view->swayc->instructions->items[i];
+		if (!instruction->ready && instruction->state.view_width == width &&
```

Comment:
What if a client "skips configures", ie. receives one configure with size WxH, and then immediately receives another configure with size W'xH', and finally decides to render one frame with size W'xH'?

Reply:
See the comment thread above this one.